subsys/usb/usb_device.c

```diff
@@ -1019,3 +1047,181 @@ int usb_ep_read_continue(u8_t ep)
 {
 	return usb_dc_ep_read_continue(ep);
 }
+
+/* Transfer management */
+static struct usb_transfer_data *usb_ep_get_transfer(u8_t ep)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(usb_dev.transfer); i++) {
+		if (usb_dev.transfer[i].ep == ep)
+			return &usb_dev.transfer[i];
+	}
+
+	return NULL;
+}
+
+#define ARRAY_INDEX(array, elemp) (((elemp) - (array)) / sizeof((array)[0]))
+
+static void usb_transfer_done(struct usb_transfer_data *trans, int status)
+{
+	int epid = ARRAY_INDEX(usb_dev.transfer, trans);
+
+	/* restore endpoint callback */
+	usb_dc_ep_set_callback(trans->ep, usb_dev.config->endpoint[epid].ep_cb);
+
+	trans->status = status;
+	k_sem_give(&trans->sem);
+
+	if (trans->cb)
+		trans->cb(trans->ep, trans->status, trans->size);
+}
+
+static void usb_handle_transfer(u8_t ep, enum usb_dc_ep_cb_status_code status)
+{
+	struct usb_transfer_data *trans;
+	int ret;
+
+	trans = usb_ep_get_transfer(ep);
+	if (!trans) {
+		return;
+	}
+
+	if (trans->status != -EBUSY) {
+		return;
+	}
+
+	if (status == USB_DC_EP_DATA_IN) { /* write complete */
+		int bytes;
+
+		if (!trans->bsize) { /* transfer complete */
+			if (trans->zero)
+				usb_dc_ep_write(ep, NULL, 0, NULL);
+			usb_transfer_done(trans, 0);
+			return;
+		}
+
+		ret = usb_dc_ep_write(ep, trans->buffer, trans->bsize, &bytes);
+		if (ret) {
+			/* transfer error */
+			usb_transfer_done(trans, -EINVAL);
+			return;
+		}
+
+		trans->buffer += bytes;
+		trans->bsize -= bytes;
+		trans->size += bytes;
+	} else if (status == USB_DC_EP_DATA_OUT) {
+		int bytes;
+
+		ret = usb_dc_ep_read(ep, trans->buffer, trans->size, &bytes);
+		if (ret) {
+			/* transfer error */
+			usb_transfer_done(trans, -EINVAL);
+			return;
+		}
+
+		trans->buffer += bytes;
+		trans->bsize -= bytes;
+		trans->size += bytes;
+
+		/* ZLP, short-pkt or buffer full */
+		if (!bytes || (bytes % usb_dc_ep_mps(ep)) || !trans->bsize) {
+			/* transfer complete */
+			usb_transfer_done(trans, 0);
+			return;
+		}
+	}
+}
+
+int usb_transfer(u8_t ep, u8_t *data, size_t dlen, unsigned int flags,
+		 usb_transfer_callback cb)
+{
+	struct usb_transfer_data *trans = usb_ep_get_transfer(ep);
+	unsigned int key;
+	int ret;
+
+	SYS_LOG_DBG("ep=%02x, data=%p, dlen=%d\n", ep, data, dlen);
+
+	if (!trans) {
+		return -EINVAL;
+	}
+
+	if (k_sem_take(&trans->sem, K_NO_WAIT)) {
+		return -EBUSY;
+	}
+
+	/* Configure transfer */
+	trans->ep = ep;
+	trans->buffer = data;
+	trans->bsize = dlen;
+	trans->size = 0;
+	trans->cb = cb;
+	trans->status = -EBUSY;
+
+	/* we need to temporaly change the ep callback */
+	usb_dc_ep_set_callback(ep, usb_handle_transfer);
+
+	key = irq_lock();
```

Comment:
I have some concerns of using the IRQ locking here. Please correct me if my understanding is incorrect.

As per my understanding, when it comes to usb_dc_ep_write(), I think the current drivers (both DW and STM32) use a straight forward method of writing into the HW registers word by word and so things are very much synchronous and simple. I'm currently developing a driver for Nordic's nRF52840 which has a partially offloaded architecture that uses DMA to copy to/from HW local buffer and then HW takes care all the read and write operations. For example, the DMA copy done and successful USB write on the bus are indicated as events through interrupts.

If we go by the current method having the IRQ locked, then I have to disable the respective interrupts and implement polling mechanism to wait for the events instead of handling them from interrupt context - making the whole thing synchronous defeating the purpose of having the DMA and HW offload advantages.

Could you please take the DMA and partial HW offload architecture use cases into consideration as well?

Reply:
If `usb_handle_transfer()` is supposed to be a synchronous function then I am not convinced this is a good idea, doing a full transfer of bytes with interrupts disabled.

Reply:
In my experiments with BT-USB so far, I have found irq_lock() has been causing packet drops and even caused asserts in the BT link layer. I had to modify my code to lock by disabling only USB peripheral interrupts. That way the critical code within the USB driver is still safe to execute while allowing other interrupts to happen elsewhere.

I have a little thought here. Can we have a lock for USB subsystem alone as a driver API so that the USB stack can make use of it?

For example, we can call the locking APIs as usb_dc_irq_enable() and usb_dc_irq_disable().
In most architectures we have a dedicated IRQ line for USB device controller that can be enabled or disabled as and when it's required.

void usb_dc_irq_enable(void)
{
        irq_enable(ARCH_SPECIFIC_IRQ_LINE);
}

void usb_dc_irq_disable(void)
{
        irq_disable(ARCH_SPECIFIC_IRQ_LINE);
}

But these too should only be used for short period of time from the USB stack, not during the whole write() operation.

Reply:
Also to change the callback without irq_lock could go wrong, line 1163. @loicpoulain maybe you can make it simpler, at least for now. Do not change the endpoint callback here, but set it directly to usb_handle_transfer in your class for example when initializing it. Change the function usb_transfer so that usb_transfer_data can be used as parameter and call usb_dc_ep_write first time from usb_transfer. Then IN transfer should go without irq_lock (I think so :smile:). OUT is not that easy, assume that the host can start a transfer every time after enumeration. You can initialize the OUT transfer from usb_handle_transfer. I suspect that it can not be so generic, do you have a branch with the example where you use it?

Reply:
@sundar
> I'm currently developing a driver for Nordic's nRF52840 which has a partially offloaded architecture that uses DMA to copy to/from HW local buffer and then HW takes care all the read and write operations. For example, the DMA copy done and successful USB write on the bus are indicated as events through interrupts.

If I understand correctly, your dc_ep_write and dc_ep_read functions are blocking operations => so they can not be executed in atomic/IRQ context, correct ? What about usb_ep_callback, is it call from IRQ or thread context ? To be honest I would expect usb_write 'atomic' and fully asynchronous since the usb_ep_callback is called on completion.

> If usb_handle_transfer() is supposed to be a synchronous function then I am not convinced this is a good idea, doing a full transfer of bytes with interrupts disabled. Disabling interrupts for an extended period of time is a serious issue for the hard real-time components of the RTOS, such as the BLE Link Layer when combined with the USB driver in order to provide a "USB Bluetooth Dongle" build.

Here I assume (maybe wrongly) that ep_read/ep_write are short operations.
usb_dc_ep_set_callback (usb_handle_transfer) is called from IRQ context (at least with stm32 and dw drivers). We can not offload this to a thread/bottom-half cause usb_dc_ep_read needs to be done into the callback itself (data is lost after that).

> @loicpoulain maybe you can make it simpler, at least for now. Do not change the endpoint callback here, but set it directly to usb_handle_transfer in your class for example when initializing it.

Good idea.

> OUT is not that easy, assume that the host can start a transfer every time after enumeration. You can initialize the OUT transfer from usb_handle_transfer. I suspect that it can not be so generic, do you have a branch with the example where you use it?

Once class driver is ready (e.g. after initialization), a read transfer can be triggered. I can prepare a branch or add example to this pull request.

Reply:
@loicpoulain 

> If I understand correctly, your dc_ep_write and dc_ep_read functions are blocking operations => so they can not be executed in atomic/IRQ context, correct ? What about usb_ep_callback, is it call from IRQ or thread context ? To be honest I would expect usb_write 'atomic' and fully asynchronous since the usb_ep_callback is called on completion.

In nRF52840, the usb_dc_ep_write() usb_dc_ep_read() functions are *not* blocking operations. They are fully asynchronous and would call the usb_ep_callback after write is complete in case of write operation.
The callback would be called from thread context because we don't want huge data processing to happen in interrupt context as this would affect further interrupts to be held waiting at the HW. Application could do anything in its write or read callback handler.

Our driver's write() does something like this:
a) Prepares the DMA engine to copy from RAM to HW's local buffer (with USBD interrupts disabled)
b) Initiate the IN transaction (with USBD interrupts disabled)
c) Return to caller (USBD interrupts are enabled)
d) USBD interrupt indicates the DMA completion followed by write completion. Driver handles them in worker thread context.
e) Worker thread calls the write callback to indicate write complete.

So, the irq_lock() added before calling usb_handler_transfer will affect the interrupt handling at the driver layer.